feat: Implement the scheduler

isidentical opened the pull request:
Depends on #406

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+    if isinstance(node, nx.DiGraph):
+        model_ids = sorted(
+            list(node),
+            key=lambda node: node == properties["exit_node"],
```
add a comment about sorting by boolean

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+
+    if isinstance(task, FalModelTask):
+        assert isinstance(model_node, DbtModelNode)
+        task.bound_model = model_node.model
```
couldn't this happen in the constructor?

isidentical replied:
Ah seems like I can simply lift the `model_node` retrieval to a bit above, seems to be solved now!

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+        self._counter += 1
+        target_node.set_run_index(self._counter)
+
+        self.nodes.remove(target_node)
+        self._staged_nodes.append(target_node)
```
Is it possible that a race condition could tink a dependency is unblocked? Should we swap the order of these 2?

isidentical replied:
Same as question as below.

chamini2 replied:
I see now that `_staged_nodes` is not really used to check if a node is in process or not. `_staged_nodes` really does not do much, we could not have it at all. Unless it is useful for debugging, maybe.

isidentical replied:
Yep, that is the primary use case (a bit generalized, like not making a node disappear all together until it is finished).

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+        # alltogether and unblock all of its dependencies.
+        self._staged_nodes.remove(target_node)
+
+        if status == SUCCESS:
+            self._succeed(target_node)
+        else:
+            self._fail(target_node)
```
should we do this before `self._staged_nodes.remove(target_node)` for race conditions?

isidentical replied:
I am unsure about how a race can happen here. Would you mind giving an example?

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+            if target_node in node.dependencies:
+                node.dependencies.remove(target_node)
+
+    def iter_available_nodes(self) -> Iterator[Node]:
```
Should this return a node instead of being an iterator of nodes? If it has a lot of unblocked nodes at the beginning without any depth, it will still process all of those.

```mermaid
graph LR;
  a;
  b --> c --> d --> e --> f --> g;
```

Would yield `b`, then `a`, and then need to be called again and get `c`, etc.

isidentical replied:
I initially did it that way, but when I wrote the executor I've noticed that it can take as many as available so instead of calculating unblocked nodes at each call we only do calculate once and give up everything we can at the same time. If we did handle the number of jobs in the queue for the thread pool executor manually, I think your suggestion might be a decent optimization but currently it is all managed by the `concurrent.futures` so even if make this a method we'll still take all the unblocked nodes at each iteration and pass them at the same time.

isidentical replied:
Maybe this could be a feature in the future though, with the improvements in the node scoring system (like timing info processing) it might matter a lot more. But for now, I'd abstain from that. WDYT?

isidentical replied:
> I say change this not to a list of nodes but to return the most scored node and that is it.

But it really doesn't solve the problem with this system. If we change this to something that only returns a single node, then we'll port this logic back to the executor which will call this method as much as it can since we don't know how many available threads are there:

```py
while (node := scheduler.get_node()) is not None:
     submit_future_group(node)
```

which would be pretty much the same with the current approach, even a bit slower and a bit more complicated. Though maybe I am missing something.

chamini2 replied:
> then we'll port this logic back to the executor which will call this method as much as it can since we don't know how many available threads are there:
> ```
> while (node := scheduler.get_node()) is not None:
>      submit_future_group(node)
> ```
> which would be pretty much the same with the current approach, even a bit slower and a bit more complicated. Though maybe I am missing something.

Got it. I see your point! I think maybe a future change would be to make the executor aware of available threads, so it just calls when it makes sense. For now, let's leave it like it is!

Thanks for the explanation!

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+
+        target_node.exit(status)
+
+    def _fail(self, target_node: Node) -> None:
+        for node in self.nodes.copy():
+            if target_node in node.dependencies:
+                self.nodes.remove(node)
```
I see dependencies have all ancestors, not just direct dependencies.

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+            yield node
+
+
+def schedule_graph(graph: nx.DiGraph, node_graph: NodeGraph) -> NodeQueue:
```
`node_graph` has a `nx.DiGraph` in `node_graph.graph`. Is the one received as first arg the same?

turbo1912 commented on src/fal/planner/tasks.py:
```diff
@@ -0,0 +1,155 @@
+
+
+@dataclass
+class Node:
```
Is `Node` the combination of a `DbtTask` and its post hooks? Seems like we need to name this something more specific

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,132 @@
+    def finish(self, target_node: Node, status: int) -> None:
+        # When a staged node's execution is finished, we'll remove it
+        # alltogether and unblock all of its dependencies.
+        self._staged_nodes.remove(target_node)
```
`_staged_nodes` really does not block stuff from starting. It is a way to keep the running nodes somewhere, but we could do without it.

isidentical commented:
@turbo1912 @chamini2 short update:

- Renamed `Node` to `TaskGroup` based on your feedback. I think that kind of covers its purpose (like multiple tasks grouped together).
- Renamed `NodeQueue` to `Scheduler` (and removed references to `node` name under it, so renamed a couple other methods).
- Added more explanation behind `exit_node` into the scheduler.
- Removed `_staged_nodes`, it didn't provide enough value to stay. So based on your suggestions I also removed it.

chamini2 commented on src/fal/planner/schedule.py:
```diff
@@ -0,0 +1,133 @@
+        else:
+            self._fail(target_group)
+
+        target_group.exit(status)
```
Right now this stops the whole execution, right?

isidentical replied:
Yep, we have [FEA-207](https://linear.app/features-and-labels/issue/FEA-207/implement-error-recovery-when-a-node-fails-or-independent-nodes-still) as a sub-issue specifically for the feature we talked.
